### PR TITLE
RSDK-4625 - getWaypointLoop should not close out forever if there are no waypoints

### PR DESCRIPTION
**Pull request by nfranczak**
this PR also subsumes RSDK-4624 and RSDK-4625

### CODE REVIEW

File: services/navigation/builtin/builtin.go
Comment (biotinker) [inline services/navigation/builtin/builtin.go:412]: A `continue` should work here, no need for loop labels.

File: services/navigation/builtin/builtin.go
Comment (biotinker) [inline services/navigation/builtin/builtin.go:1]: There are going to be a bunch of tests that will need to be updated with this change as well I suspect.

File: services/navigation/builtin/builtin.go
Comment (nfranczak) [inline services/navigation/builtin/builtin.go:411]: i have to do `err != nil` first and have this conditional use the logical and operator since we want to avoid "nil pointer dereference"

Comment (biotinker) [inline services/navigation/builtin/builtin.go:411]: This should be compared to a typed error, rather than a raw string comparison.

**Review by biotinker — CHANGES_REQUESTED** [review]

File: services/navigation/builtin/builtin.go
Comment (nfranczak) [reply services/navigation/builtin/builtin.go:412]: I disagree that `continue` would be suffice.
My reasoning is:
suppose we have no remaining waypoints to navigate to.
the `NextWaypoint` method will then return `Waypoint{}, errNoMoreWaypoints`
if the program is allowed to `continue` we will eventually reach `navOnce` which expects a  context + waypoint
then inside `navOnce` we would do `wp.ToPoint(),` to give the destination to the moveonglobe call.
since `wp := Waypoint{}`, then `wp.ToPoint() = 0,0`
we do not want to navigate 0 0

Comment (biotinker) [reply services/navigation/builtin/builtin.go:412]: In order for that to happen, we would have to pass this check, so at that point `wp` would be valid.

`continue` will jump to a new iteration of the innermost loop. It doesn't mean "continue with the next line".

File: web/frontend/src/components/navigation/index.svelte
Comment (micheal-parks) [inline web/frontend/src/components/navigation/index.svelte:9]: Since we're only using `navMode` in this file we can do away with stores, in svelte variables themselves trigger reactive changes in html if they're updated, so something like this should work:

```svelte
<script lang='ts'>
  let navMode: 'Manual' | 'Waypoint' | '' = ''
  
  // ...later
  navMode = 'Manual'
<script>

<v-radio
  label="Navigation mode"
  options="Manual, Waypoint"
  selected={navMode}
  on:input={setNavigationMode}
/>

File: web/frontend/src/components/navigation/index.svelte
Comment (micheal-parks) [inline web/frontend/src/components/navigation/index.svelte:29]: can we make some enums for these number values?

```ts
const enum NavigationModes {
  Manual = 1,
  Waypoint = 2,
}

File: services/navigation/builtin/builtin.go
Comment (nfranczak) [reply services/navigation/builtin/builtin.go:412]: > It doesn't mean "continue with the next line".

ai yayay. you're right, I'm sorry - not sure what I was thinking 🧠 💨

File: services/navigation/builtin/builtin.go
Comment (raybjork) [inline services/navigation/builtin/builtin.go:413]: Not sure if we should even be exiting here.  If we have an error from this function but we can't return it, it will quietly shut down the navigation service loop.  I'm thinking instead we should be just ignoring the error and continuing.  This would mean we wouldn't need to check the type of the error

File: services/navigation/builtin/builtin.go
Comment (raybjork) [reply services/navigation/builtin/builtin.go:1]: To add to this, I would be interested to see that this loop properly exits when we switch from waypoint to manual mode.  (this might already be captured in a test I didn't look very thoroughly)

**Review by raybjork — COMMENTED** [review]
Looks good, Peter covered all my concerns.  Just have one question about what desired behavior should be

File: services/navigation/builtin/builtin.go
Comment (nfranczak) [inline services/navigation/builtin/builtin.go:280]: it strikes me as strange that this how the original code was set-up . is this intentional. 
from my POV this replacement seems benign, but I want to double check.

Comment (biotinker) [reply services/navigation/builtin/builtin.go:280]: This is intentional and correct; you'll want to undo this change.

But let's add a comment explaining why.

The `ctx` passed in has a lifespan that only runs as long as the duration of the call. So since this ctx comes from the `SetMode` call, it would be cancelled (and end this loop) once SetMode finishes setting the mode. This isn't what we want; we want it to loop indefinitely in the background until it encounters a specific error condition. So `context.Background()` is what we want here.

**Review by micheal-parks — APPROVED** [review]
Only reviewed the client side, which looks good 👍

**Review by nicksanford — COMMENTED** [review]
This seems reasonable to me but I'm still ramping so not approving so that you get approval from another team member before merging.

File: services/navigation/builtin/builtin_test.go
Comment (raybjork) [inline services/navigation/builtin/builtin_test.go:211]: this looks really similar to the above function, maybe we could collapse them into one to avoid adding reduntant tests

**Review by raybjork — APPROVED** [review]
I have one comment about testing practices that I would like addressed but other than that LGTM

**Review by biotinker — APPROVED** [review]
LGTM

File: services/navigation/builtin/builtin_test.go
Comment (nfranczak) [reply services/navigation/builtin/builtin_test.go:211]: I agree it is super similar. I took a thorough look through the tests in this file and I believe that this test is actually a duplicate of one which already exists.
```
t.Run("Calling SetMode cancels current and future MoveOnGlobe calls", func(t *testing.T) {
```
for this reason, i will just remove this sub-test.

**Pull request merged**
